src/common/os_dimm_windows.c

```diff
@@ -0,0 +1,174 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * os_dimm_none.c -- fake dimm stats
+ */
+
+#include "out.h"
+#include "os.h"
+#include "os_dimm.h"
+#include "util.h"
+
+#define GUID_SIZE 37 /* "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX\0" */
+
+/*
+ * os_dimm_volume -- returns volume handle
+ */
+static HANDLE
+os_dimm_volume_handle(const char *path)
+{
+	char drive[3]; /* Drive letter + ':' + '\0' */
+	_splitpath(path, drive, NULL, NULL, NULL);
+	char vol[7]; /* "\\\\.\\X:\0" */
+	sprintf(vol, "\\\\.\\%s", drive);
+	return CreateFileA(vol, 0, FILE_SHARE_READ | FILE_SHARE_WRITE,
+		NULL, OPEN_EXISTING, 0,  NULL);
+}
+
+/*
+ * os_dimm_uid -- returns a file uid based on dimm guid
+ */
+int
+os_dimm_uid(const char *path, char *uid, size_t *len)
+{
+	LOG(3, "path %s, uid %p, len %lu", path, uid, *len);
+	if (uid == NULL) {
+		*len = GUID_SIZE;
+	} else {
+		ASSERT(*len >= GUID_SIZE);
+		HANDLE vHandle = os_dimm_volume_handle(path);
+		if (vHandle == INVALID_HANDLE_VALUE)
+			return -1;
+
+		STORAGE_DEVICE_NUMBER_EX sdn;
+		sdn.DeviceNumber = -1;
+		DWORD dwBytesReturned = 0;
+		if (!DeviceIoControl(vHandle,
+				IOCTL_STORAGE_GET_DEVICE_NUMBER_EX,
+				NULL, 0,
+				&sdn, sizeof(sdn),
+				&dwBytesReturned, NULL)) {
+			/*
+			 * IOCTL_STORAGE_GET_DEVICE_NUMBER_EX is not supported
+			 * on this server
+			 */
+			memset(uid, 0, *len);
+			uid = ZERO_GUID;
+			CloseHandle(vHandle);
+			return 0;
+		}
+		GUID guid = sdn.DeviceGuid;
+		sprintf(uid,
+			"%08lX-%04hX-%04hX-%02hhX%02hhX-%02hhX%02hhX%02hhX%02hhX%02hhX%02hhX",
+			guid.Data1, guid.Data2, guid.Data3, guid.Data4[0],
+			guid.Data4[1], guid.Data4[2], guid.Data4[3],
+			guid.Data4[4], guid.Data4[5], guid.Data4[6],
+			guid.Data4[7]);
+		CloseHandle(vHandle);
+	}
+	return 0;
+}
+
+/*
+ * os_dimm_usc -- returns unsafe shutdown count
+ */
+int
+os_dimm_usc(const char *path, uint64_t *usc)
+{
+	LOG(3, "path %s, usc %p", path, usc);
+	*usc = 0;
+	HANDLE vHandle = os_dimm_volume_handle(path);
+	if (vHandle == INVALID_HANDLE_VALUE)
+		return -1;
+
+	STORAGE_PROPERTY_QUERY prop;
+	DWORD dwSize;
+	prop.PropertyId = StorageDeviceUnsafeShutdownCount;
+	prop.QueryType = PropertyExistsQuery;
+	STORAGE_DEVICE_UNSAFE_SHUTDOWN_COUNT ret;
+
+	BOOL bResult = DeviceIoControl(vHandle,
+		IOCTL_STORAGE_QUERY_PROPERTY,
+		&prop, sizeof(STORAGE_PROPERTY_QUERY),
+		&ret, sizeof(STORAGE_DEVICE_UNSAFE_SHUTDOWN_COUNT),
+		(LPDWORD)&dwSize, (LPOVERLAPPED)NULL);
+
+	if (!bResult) {
+		CloseHandle(vHandle);
+		return 0; /* STORAGE_PROPERTY_QUERY not supported */
+	}
+	prop.QueryType = PropertyStandardQuery;
+	bResult = DeviceIoControl(vHandle,
```

Comment:
Is it required by winapi to send the ioctl twice, first to verify if some property exists and then to actually read that property?